src/engine/plugins/match.cpp

```diff
@@ -234,12 +265,25 @@ Status MatchPlugin::HandleRequest(const RoutingAlgorithmsInterface &algorithms,
             BOOST_ASSERT(current_phantom_node_pair.target_phantom.IsValid());
             sub_routes[index].segment_end_coordinates.emplace_back(current_phantom_node_pair);
         }
-        // force uturns to be on, since we split the phantom nodes anyway and only have
-        // bi-directional
-        // phantom nodes for possible uturns
+        // force uturns to be on
+        // we split the phantom nodes anyway and only have bi-directional phantom nodes for
+        // possible uturns
         sub_routes[index] =
             algorithms.ShortestPathSearch(sub_routes[index].segment_end_coordinates, {false});
         BOOST_ASSERT(sub_routes[index].shortest_path_weight != INVALID_EDGE_WEIGHT);
+        if (!tidied.parameters.waypoints.empty())
+        {
+            std::vector<bool> waypoint_legs;
+            waypoint_legs.reserve(sub_matchings[index].nodes.size());
+            for (unsigned i = 0; i < sub_matchings[index].nodes.size(); ++i)
```

Comment:
I think we can do this in a single sweep over the indices of the submatch:
- Allocate a vector of size #waypoints 
- Iterate over indices and `parameters.waypoints` at the same time (they are both sorted increasing)
- Set `waypoints_legs[i]` to true if we find a match.